smart_open/s3.py

```diff
@@ -1189,18 +1189,31 @@ def iter_bucket(
 
     with smart_open.concurrency.create_pool(processes=workers) as pool:
         result_iterator = pool.imap_unordered(download_key, key_iterator)
-        for key_no, (key, content) in enumerate(result_iterator):
-            if True or key_no % 1000 == 0:
-                logger.info(
-                    "yielding key #%i: %s, size %i (total %.1fMB)",
-                    key_no, key, len(content), total_size / 1024.0 ** 2
-                )
-            yield key, content
-            total_size += len(content)
-
-            if key_limit is not None and key_no + 1 >= key_limit:
-                # we were asked to output only a limited number of keys => we're done
+        key_no = 0
+        while True:
+            try:
+                (key, content) = result_iterator.__next__()
+                if True or key_no % 1000 == 0:
+                    logger.info(
+                        "yielding key #%i: %s, size %i (total %.1fMB)",
+                        key_no, key, len(content), total_size / 1024.0 ** 2
+                    )
+                yield key, content
+                total_size += len(content)
+                if key_limit is not None and key_no + 1 >= key_limit:
+                    # we were asked to output only a limited number of keys => we're done
+                    break
+            except botocore.exceptions.ClientError as err:
+                if 'Error' in err.response and err.response['Error'].get('Code') == '404':
+                    logger.warning(
+                        "Encountered '404 Not Found' error for key #%i",
+                        key_no
```

Comment:
Would it perhaps make sense to log the key, rather than the key_no? If anybody does care to see the log, I'm not sure the number would be as helpful in debugging a situation.

An argument could perhaps be made that it's not actually worthy of being logged as a warning. An end-user could read this as something being wrong or broken, when this is completely valid behavior that is to be expected from time to time.

Reply:
That is a very good point. If a 404 error occurs during iter bucket, it is almost certainly not the users fault, but rather the fault of the internal list keys function. I agree that logging the error would just be confusing to the user. I've removed the warning in my latest commit. Let me know what you think.

Reply:
Makes sense to me 👍